In [1]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [2]:
# Load the given datasets
categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")


# Print the data
print(categorical)
print(numerical)



     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

Features are the input variables or attributes used for prediction in a model. The dataset contains the following features:
age: The patient's age (numerical)
bp: Blood pressure (numerical)
sg: Specific gravity (numerical)
al: Albumin (categorical: presence or absence of albumin in urine)
su: Sugar (categorical: presence or absence of sugar in urine)
rbc: Red blood cells (categorical: presence or absence of RBCs in urine)
pc: Pus cell (categorical: presence or absence of pus cells in urine)
pcc: Pus cell clumps (categorical: presence or absence of pus cell clumps)
ba: Bacteria (categorical: presence or absence of bacteria in urine)
bgr: Blood glucose random (numerical)
bu: Blood urea (numerical)
sc: Serum creatinine (numerical)
sod: Sodium (numerical)
pot: Potassium (numerical)
hemo: Hemoglobin (numerical)
pcv: Packed cell volume (numerical)
wc: White blood cell count (numerical)
rc: Red blood cell count (numerical)
htn: Hypertension (categorical: indicates if the patient has hypertension)
dm: Diabetes mellitus (categorical: indicates if the patient has diabetes)
cad: Coronary artery disease (categorical: indicates if the patient has CAD)
appet: Appetite (categorical: indicates if the patient has a normal or poor appetite)
pe: Pedal edema (categorical: indicates if the patient has pedal edema)
ane: Anemia (categorical: indicates if the patient has anemia)

- The features in the given datasets are both numerical and categorical.

Categorical Features
Categorical variables represent categories or groups and typically include non-numeric data. Examples of categorical features from this dataset:
al: Albumin presence (yes/no)
rbc: Red blood cell presence (yes/no)
htn: Hypertension status (yes/no)

Numerical Features
Numerical variables represent quantifiable measurements and can be discrete or continuous. Examples of numerical features from this dataset:
age: Age of the patient
bp: Blood pressure measurement
bgr: Random blood glucose level

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
nDuplicates = numerical.duplicated(subset=['unique_id'])
nNumDupes = nDuplicates.sum()
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

if (nNumDupes > 0):
    print(nNumDupes)


# Drop these duplicate rows

numerical.drop_duplicates(subset=['unique_id'], inplace=True)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them

cDuplicates = categorical.duplicated(subset=['unique_id'])
cNumDupes = cDuplicates.sum()

if(cNumDupes > 0):
    print(cNumDupes)

categorical.drop_duplicates(subset=['unique_id'], inplace=True)



30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

No, all the unique ids are not present in both datasets. This could be becuase of errors made while entering in data. The rows that are missing from one of the datasets but prsent in the other are present in the combined table because of the way merge works.

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

combined = pd.merge(numerical, categorical, on='unique_id', how='outer')

#Print the combined dataset

print(combined)


     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value

numMissing = combined.isnull().any(axis=1).sum()

percent = numMissing/(len(combined))

# Print %

print(percent*100)

# Drop these rows from the dataset

combined.dropna(inplace=True)

# Print the Dataset

print(combined)

61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc        

## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

combined.sort_values(by='Target', inplace=True)
combined.reset_index(drop=True, inplace=True)


# Print the dataset
print(combined)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

withDummies = pd.get_dummies(combined, columns=categorical_columns, drop_first=True)

# Print the dataset

print(withDummies)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [8]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for column in numerical_columns:
    mean = withDummies[column].mean() #might have to change to combined instead?
    std_dev = withDummies[column].std()
    
    # Define the boundaries for outliers
    upper_limit = mean + 3 * std_dev
    lower_limit = mean - 3 * std_dev

    withDummies = withDummies[(withDummies[column] <= upper_limit) & (withDummies[column] >= lower_limit)]
   
# Print the dataset

print(withDummies)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [9]:
# Normalize the all Numerical Attributes in the dataset.

withDummies[numerical_columns] = (withDummies[numerical_columns] - withDummies[numerical_columns].min()) / (withDummies[numerical_columns].max() - withDummies[numerical_columns].min())

# Print the dataset

print(withDummies)

     unique_id       age    bp       bgr        bu        sc       sod  \
2       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
5       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
7       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
8       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
9       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
148     486397  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
149     203466  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
150     482105  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
151     228849  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
152     996076  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv  ...  pc_normal  pcc_present  ba_present  \
2    0.793103  0.000000  0.032258

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

Yes, age is not appropriate for modeling and predictions. I removed it because it doesn't provide meaningful information for predicting CKD and could make the model more complicated and less fair. It doesn't seem to have much value when it comes to modeling and predicting, also if the ages are too similar then it might not be useful to include.

In [10]:
#Remove that column

withDummies.drop('age', axis=1, inplace=True)
# Print the dataset

print(withDummies)

     unique_id    bp       bgr        bu        sc       sod       pot  \
2       343710  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5       397388  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7       438182  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8       474407  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9       475200  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
..         ...   ...       ...       ...       ...       ...       ...   
148     486397  0.25  0.227468  0.222222  0.074074  0.500000  0.310345   
149     203466  0.00  0.115880  0.052288  0.098765  0.600000  0.482759   
150     482105  0.00  0.107296  0.235294  0.012346  0.600000  0.448276   
151     228849  0.50  0.128755  0.261438  0.098765  0.666667  0.206897   
152     996076  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc  ...  pc_normal  pcc_present  ba_present  \
2    0.000000  0.032258  0.395161

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [11]:
# Export the dataframe to a new csv file

temp = withDummies

temp.to_csv('new_data_csv.csv', index=False, header=True, encoding='utf-8')

# Export the dataframe to a new json file

temp.to_json('new_data_json.json', index=False)

# print(temp.head(15))

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here
| unique_id |   bp |      bgr |       bu |       sc |       sod |       pot |
|-----------|------|----------|----------|----------|-----------|-----------|
|     343710 | 0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000  | 0.793103  |
|     397388 | 1.00 | 0.901288 | 0.163399 | 0.345679 | 0.766667  | 0.206897  |
|     438182 | 0.50 | 0.785408 | 0.862745 | 0.518519 | 0.600000  | 1.000000  |
|     474407 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667  | 0.206897  |
|     475200 | 1.00 | 0.399142 | 0.287582 | 0.839506 | 0.666667  | 0.586207  |
|     481293 | 0.00 | 0.935622 | 0.169935 | 0.160494 | 0.333333  | 0.034483  |
|     514721 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667  | 0.655172  |
|     995177 | 0.25 | 0.618026 | 0.562092 | 0.728395 | 0.000000  | 0.344828  |
|     546225 | 0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000  | 0.379310  |
|     980291 | 0.25 | 0.223176 | 0.209150 | 0.160494 | 0.533333  | 0.620690  |
|     955830 | 0.50 | 0.618026 | 0.411765 | 0.432099 | 0.566667  | 0.689655  |
|     923613 | 0.00 | 0.725322 | 0.313725 | 0.481481 | 0.566667  | 0.862069  |
|     843362 | 0.00 | 0.206009 | 0.751634 | 0.604938 | 0.533333  | 0.689655  |
|     828592 | 0.25 | 0.639485 | 0.470588 | 0.395062 | 0.433333  | 0.517241  |
|     621332 | 0.50 | 1.000000 | 0.163399 | 0.111111 | 0.066667  | 0.206897  |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [12]:
# Code to rename all the columns in the dataset
column_mappings = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Step 1: Rename the columns in the DataFrame
temp.rename(columns=column_mappings, inplace=True)

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [14]:
# Converted SQL to Pandas code
result = combined.groupby('Target').size().reset_index(name='count')
print(result)

   Target  count
0     ckd     42
1  notckd    111


*Describe what the above code does here*

The code first groups the data in the adaset by the Target column. Then, size() counts the # of entries in each group created by the groupB operation. This gives us the total count of occurances for each unique value in the Target column. Afterwards, it converts the grouped data back into a dataframe and creates a new column named count that contains teh counts from the previous steps. The reset_index() also resets the index of dataframe. Then we print out the result which will be the count of ckd and the count for notckd.